pkg/proxy/proxy.go

```diff
@@ -264,51 +266,214 @@ func (p *Proxy) health(ctx echo.Context) error {
 	requestedNamespace := namespaceFromCtx(ctx)
-	if err := validateWorkspaceRequest(workspaceName, requestedNamespace, workspaces); err != nil {
-		return "", nil, crterrors.NewForbiddenError("invalid workspace request", err.Error())
+	if err := validateWorkspaceRequest("", requestedNamespace, workspaces...); err != nil {
+		return nil, crterrors.NewForbiddenError("invalid workspace request", err.Error())
 	}
 
-	return proxyPluginName, cluster, nil
+	// return the cluster access
+	return cluster, nil
+}
+
+// processWorkspaceRequest process an HTTP Request targeting a specific workspace.
+func (p *Proxy) processWorkspaceRequest(ctx echo.Context, userID, username, workspaceName, proxyPluginName string) (*access.ClusterAccess, error) {
+	// check that the user is provisioned and the space exists
+	if err := p.checkUserIsProvisionedAndSpaceExists(ctx, userID, username, workspaceName); err != nil {
+		return nil, err
+	}
+
+	// retrieve the requested Workspace with SpaceBindings
+	workspace, err := p.getUserWorkspaceWithBindings(ctx, workspaceName)
+	if err != nil {
+		return nil, err
+	}
+
+	// check whether the user has access to the workspace
+	// and whether the requestedNamespace -if any- exists in the workspace.
+	requestedNamespace := namespaceFromCtx(ctx)
+	if err := validateWorkspaceRequest(workspaceName, requestedNamespace, *workspace); err != nil {
+		return nil, crterrors.NewForbiddenError("invalid workspace request", err.Error())
+	}
+
+	// retrieve the ClusterAccess for the user and the target workspace
+	return p.getClusterAccess(ctx, userID, username, workspaceName, proxyPluginName, workspace)
+}
+
+// checkUserIsProvisionedAndSpaceExists checks that the user is provisioned and the Space exists.
+// If the PublicViewer support is enabled, User check is skipped.
+func (p *Proxy) checkUserIsProvisionedAndSpaceExists(ctx echo.Context, userID, username, workspaceName string) error {
+	if err := p.checkUserIsProvisioned(ctx, userID, username); err != nil {
+		return crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
+	}
+	if err := p.checkSpaceExists(workspaceName); err != nil {
+		return crterrors.NewInternalError(errs.New("unable to get target cluster"), err.Error())
+	}
+	return nil
+}
+
+// checkSpaceExists checks whether the Space exists.
+func (p *Proxy) checkSpaceExists(workspaceName string) error {
+	if _, err := p.app.InformerService().GetSpace(workspaceName); err != nil {
+		// log the actual error but do not return it so that it doesn't reveal information about a space that may not belong to the requestor
+		log.Error(nil, err, "unable to get target cluster for workspace "+workspaceName)
+		return fmt.Errorf("the requested space is not available")
+	}
+	return nil
+}
+
+// checkUserIsProvisioned checks whether the user is Approved, if they are not an error is returned.
+// If public-viewer is enabled, user validation is skipped.
```

Comment:
why skipping the validation if public-viewer is enabled ? I mean should we skip it if both following conditions are true:
- public viewer functionality is enabled
- the requested workspace is a community one

Reply:
Also I think this will fail the ban usersignup e2e tests, since it will not block requests from banned users if I'm not mistaken. I'm planning to try and add a unit test for it as well, as I see that is not covered here.

Reply:
with public-viewer enabled we just need a valid SSO Token, we don't require the user to have Signed Up

Reply:
right, but only if the request was meant for accessing a community workspace, and not for any workspace right ? But I'm sure I'm missing something.

Reply:
Whether the user has access to this workspace or not is checked later in the chain. This method only checks that the user is provisioned and it only applies to the scenario in which Public-Viewer is disabled.

The user access check is performed by retrieving the requested workspace with SpaceBindings (cf. L332) and retrieving the Cluster Access (cf L345, L409, L420)

Reply:
Regarding banned user, I think you may be correct, I'll double check. Great point!